File: src/core/session_label.py
"""The user-facing LABEL, and the lossy filter that derives a tmux name.

TWO THINGS THAT WERE ONE THING. A session's displayed name and its tmux
session name were the same string, so renaming a session meant
``tmux rename-session``, which moves the one field the identity key
``(socket, name, epoch)`` was built on. The stored row stopped matching
the live listing, the reaper called it dead, and the same live session
came back through the adopt path as a stranger and got a second row. One
session, two rows, one of them a corpse.

Separating them does not patch that path - it makes it UNREACHABLE from
the rename surface. A label lives in ``sessions.title``. Changing it
writes one column. The tmux name and the creation epoch never move, so
the identity triple is never touched by a rename at all.

It also removes a restriction the user asked to have removed: the old
rename endpoint enforced ``^[A-Za-z0-9_-]{1,64}$``, so "Media
Compression" was not a legal session name. A label is a label; it takes
what a human types.

WHAT WAS MEASURED ABOUT TMUX, RATHER THAN ASSUMED. On a throwaway socket,
``new-session -s`` ACCEPTED every one of: a space, ``:``, ``.``, ``/``,
``"``, ``$``. So tmux is far more permissive than the old validator
implied. Two real constraints remain:

  NON-ASCII IS NOT PRESERVED. A session created as ``emoji 🚀 ok`` reads
  back from ``list-sessions`` as ``emoji __ ok``. tmux mangles it, so a
  derived name containing one would not round-trip and every lookup by
  name would miss.

  ``:`` AND ``.`` ARE TARGET SYNTAX. tmux's ``-t`` grammar splits on
  them, so a name carrying either is legal to CREATE and awkward to
  ADDRESS - which is what every other call site in this codebase does.

Neither constrains the LABEL, because the label is never handed to tmux.

THE FILTER IS LOSSY AND THEREFORE NOT AN IDENTITY. ``client: acme`` and
``client. acme`` both sanitise to ``client_acme``. That is deliberate and
safe here precisely because the tmux name no longer identifies anything
on its own - the row is identified by ``(socket, name, epoch)`` and, for
a rename, by ``(epoch, #{session_id})``. The collision is resolved by
appending a suffix, never by rejecting the user's label, and never by
letting two live sessions share a tmux name.

WHY THIS DOES NOT DISTURB SIDEBAR GROUPS. Group membership is keyed on
``tmux_name`` on purpose: an unadopted tmux session has no ``sessions``
row to key on, so a row-keyed membership could not represent it at all.
That reasoning is untouched here. The tmux name is now MORE stable than
before, not less - nothing renames it after creation - so the group key
is strictly safer than it was.
"""

from __future__ import annotations

import hashlib
import re
import sqlite3
import unicodedata
from typing import Iterable, Optional, Set

import structlog

from src.core.trail_entry import utc_now

logger = structlog.get_logger()

#: Longest label we will store. Generous, because it is a human label and
#: not a key, but bounded so no surface has to defend against an
#: unbounded string and no row can be made pathological on purpose.
LABEL_MAX_CHARS: int = 200

#: Longest derived tmux name. tmux itself imposes no such limit; this
#: keeps the value comfortable inside a FIFO filename and a status line.
TMUX_NAME_MAX_CHARS: int = 64

#: The prefix this app puts on the tmux names it creates. Stripped when
#: deriving a label from an existing name, because it is an artefact of
#: the launcher rather than anything a user typed.
APP_TMUX_PREFIX = "cloude_"

#: Used when a label sanitises down to nothing at all. tmux will not take
#: an empty name and the caller has no better fallback than we do.
FALLBACK_TMUX_NAME = "session"

#: Characters replaced by ``_`` in a derived tmux name: tmux target
#: separators plus anything whitespace-like. Applied AFTER non-ASCII has
#: been dropped, so this only ever sees a byte it can reason about.
_UNSAFE_FOR_TMUX = re.compile(r"[\s:./\\'\"`$;|&<>(){}\[\]*?!#%^~,=+@]+")

#: Control characters, which no label may contain. Includes newline and
#: tab: a label is rendered on one line everywhere it appears, and a
#: newline would either truncate it or break the surface showing it.
_CONTROL_CHARS = re.compile(r"[\x00-\x1f\x7f]")


class InvalidLabel(ValueError):
    """A label that cannot be stored or rendered as given.

    Description: raised rather than silently corrected, because quietly
      mangling what a user typed and then showing them the mangled
      version is how a UI teaches people not to trust it. The one
      correction applied without complaint is stripping surrounding
      whitespace, which is essentially always a paste artefact.
    """


def validate_label(label: Optional[str]) -> str:
    """Accept a human label, or say exactly why it cannot be one.

    Description: permissive on PURPOSE. Spaces, punctuation, mixed case
      and non-ASCII are all fine - this string is never handed to tmux,
      so tmux's constraints do not apply to it. Only two classes are
      refused: empty (there is nothing to show) and control characters
      (they break the single-line surfaces that render it).
    Inputs: label (str | None) - the raw value from the user.
    Output: str - the label to store, surrounding whitespace stripped.
    Raises: InvalidLabel - empty, too long, or carrying a control char.
    Example: validate_label('  Media Compression ')  # 'Media Compression'
    """
    if label is None:
        raise InvalidLabel("a session label cannot be empty")
    cleaned = label.strip()
    if not cleaned:
        raise InvalidLabel("a session label cannot be empty")
    if len(cleaned) > LABEL_MAX_CHARS:
        raise InvalidLabel(
            f"a session label may be at most {LABEL_MAX_CHARS} characters"
        )
    if _CONTROL_CHARS.search(cleaned):
        raise InvalidLabel(
            "a session label cannot contain control characters such as a "
            "newline or a tab"
        )
    return cleaned


def sanitize_tmux_name(label: str) -> str:
    """Derive a tmux-safe session name from a free-form label.

    Description: THE FILTER, and it is lossy by design - see the module
      docstring. Three passes, in this order:

        1. Drop everything non-ASCII. Measured: tmux does not preserve
           it, so a name carrying it would not round-trip through
           ``list-sessions`` and every lookup by name would miss.
        2. Replace runs of tmux-hostile characters with a single ``_``.
           That set leads with ``:`` and ``.`` because they are tmux's
           own ``-t`` target separators, and continues through the shell
           metacharacters, since this value ends up in command lines and
           in a FIFO filename.
        3. Trim leading and trailing ``_`` and truncate.

      Never returns an empty string: a label of ``!!!`` reduces to
      nothing and gets ``FALLBACK_TMUX_NAME`` instead, because tmux would
      refuse the empty name and this function's caller has no better
      answer available than this one does.
    Inputs: label (str) - any label; need not have been validated.
    Output: str - a non-empty ASCII name safe in a tmux target.
    Example: sanitize_tmux_name('client: acme')  # 'client_acme'
    """
    folded = unicodedata.normalize("NFKD", label or "")
    ascii_only = folded.encode("ascii", "ignore").decode("ascii")
    replaced = _UNSAFE_FOR_TMUX.sub("_", ascii_only)
    replaced = re.sub(r"_+", "_", replaced).strip("_")
    trimmed = replaced[:TMUX_NAME_MAX_CHARS].strip("_")
    return trimmed or FALLBACK_TMUX_NAME


def unique_tmux_name(label: str, *, taken: Iterable[str]) -> str:
    """Derive a tmux name that no live session already holds.

    Description: the collision half of a lossy filter. Two labels CAN
      sanitise to one name, and the answer is a suffix rather than a
      rejection: the user's label is theirs and is stored intact, while
      the tmux name is an internal handle that nobody reads and nothing
      identifies a session by on its own.

      The walk is BOUNDED. ``_2`` through ``_49``, and if every one of
      those is taken it falls back to an 8-hex-character digest of the
      label plus the taken-set size, which cannot loop. An unbounded
      "keep incrementing" would be a hang under a pathological set, and
      a hang in the create path is worse than an ugly name.
    Inputs: label (str) - the user's label. taken (Iterable[str]) - the
      names already in use on this socket.
    Output: str - a name not present in ``taken``.
    Example: unique_tmux_name('acme', taken={'acme'})  # 'acme_2'
    """
    used: Set[str] = set(taken)
    base = sanitize_tmux_name(label)
    if base not in used:
        return base
    room = max(1, TMUX_NAME_MAX_CHARS - 4)
    stem = base[:room].strip("_") or FALLBACK_TMUX_NAME
    for suffix in range(2, 50):
        candidate = f"{stem}_{suffix}"
        if candidate not in used:
            return candidate
    digest = hashlib.sha256(
        f"{label}|{len(used)}".encode("utf-8")
    ).hexdigest()[:8]
    return f"{stem[: TMUX_NAME_MAX_CHARS - 9]}_{digest}"


def label_from_tmux_name(tmux_name: Optional[str]) -> Optional[str]:
    """Derive a readable label from an existing tmux session name.

    Description: used to give rows that predate the label a sensible one
      instead of a blank. Two transforms, both reversing something the
      app did rather than something the user chose: the ``cloude_``
      prefix the launcher adds, and the underscores the filter
      substituted for spaces.

      THE PREFIX IS ONLY REMOVED WHEN SOMETHING SURVIVES IT. A name of
      exactly ``cloude_`` keeps its stem rather than becoming empty; a
      blank label is worse than an ugly one, and the whole point of this
      column is that a row always has something to show.
    Inputs: tmux_name (str | None).
    Output: str | None - the derived label, or None for no input.
    Example: label_from_tmux_name('cloude_Media')  # 'Media'
    """
    if not tmux_name:
        return None
    stem = tmux_name
    if stem.startswith(APP_TMUX_PREFIX) and len(stem) > len(APP_TMUX_PREFIX):
        stem = stem[len(APP_TMUX_PREFIX) :]
    return stem.replace("_", " ").strip() or tmux_name


def set_label(
    conn: sqlite3.Connection,
    *,
    session_uuid: str,
    label: str,
    now: Optional[str] = None,
) -> bool:
    """Store a user-chosen label on one session row.

    Description: THE USER SET, and deliberately not the same operation as
      ``session_lineage._maybe_set_title``. That one SEEDS a title and
      refuses to overwrite, because a fork's inherited title must not
      flap every time a hook fires. This one is a person deciding what
      their session is called, so it overwrites - the two are different
      operations on one column and keeping them apart is why neither has
      to compromise.

      IDENTITY IS NOT IN THE UPDATE. ``tmux_name``, ``tmux_created_epoch``
      and ``tmux_session_id`` are not in the SET clause and never will
      be. That is the property that makes a rename structurally incapable
      of splitting a session into two rows, so it is asserted by a test
      rather than left to review.

      The caller owns the transaction, matching every other write path in
      this package.
    Inputs: conn (sqlite3.Connection). session_uuid (str) - the row's
      external identity, which survives a tmux rename and is therefore
      the right handle for this. label (str) - validated here, before any
      write. now (str | None) - ISO stamp, defaults to ``utc_now()``.
    Output: bool - True when a row was updated. False means NO SUCH ROW,
      which is a definite negative the caller can render, not an error.
    Raises: InvalidLabel - the label cannot be stored; nothing is written.
    Example: set_label(conn, session_uuid='u1', label='Media Compression')
    """
    cleaned = validate_label(label)
    stamp = now or utc_now()
    cursor = conn.execute(
        "UPDATE sessions SET title = ?, updated_at = ? WHERE session_uuid = ?",
        (cleaned, stamp, session_uuid),
    )
    if cursor.rowcount > 0:
        logger.info(
            "session_label_set",
            session_uuid=session_uuid,
            note="title only; no identity column is in this statement",
        )
        return True
    logger.warning(
        "session_label_no_such_row",
        session_uuid=session_uuid,
        note="no row carries that session_uuid, so nothing was labelled",
    )
    return False


def set_label_for_instance(
    conn: sqlite3.Connection,
    *,
    socket: str,
    name: str,
    epoch: Optional[int],
    label: str,
    now: Optional[str] = None,
) -> bool:
    """Store a label on the row for one tmux INSTANCE.

    Description: the same write as :func:`set_label`, keyed on the
      instance triple instead of the row uuid, because that is what the
      HTTP rename surface has in hand - it knows which tmux session the
      user is looking at, not which row id.

      KEYED ON THE FULL TRIPLE, NEVER THE NAME ALONE. Names are reusable
      and this table can legitimately hold several rows under one name
      with different epochs (history of a name that was used twice). A
      write keyed on the name alone would relabel all of them, which is
      the same class of defect as adopting by name - and it is asserted
      by a test rather than left to review. A None epoch has no identity
      to key on and refuses immediately.

      IDENTITY IS NOT IN THE UPDATE. ``tmux_name``, ``tmux_created_epoch``
      and ``tmux_session_id`` appear only in the WHERE clause. That is
      what makes a rename structurally incapable of moving a session's
      identity, which is the entire point of separating the two.
    Inputs: conn (sqlite3.Connection). socket (str), name (str), epoch
      (int | None) - the instance triple. label (str) - validated before
      any write. now (str | None) - ISO stamp.
    Output: bool - True when a row was updated; False means NO SUCH
      INSTANCE, a definite negative rather than an error.
    Raises: InvalidLabel - nothing is written.
    Example:
        set_label_for_instance(conn, socket='cloude', name='cloude_a',
                               epoch=7, label='Media Compression')
    """
    cleaned = validate_label(label)
    if epoch is None:
        logger.warning(
            "session_label_no_epoch",
            tmux_socket=socket,
            tmux_name=name,
            note=(
                "no creation epoch, so there is no instance to key on; "
                "nothing written rather than relabelling by name"
            ),
        )
        return False
    stamp = now or utc_now()
    cursor = conn.execute(
        "UPDATE sessions SET title = ?, updated_at = ? "
        "WHERE tmux_socket = ? AND tmux_name = ? AND tmux_created_epoch = ?",
        (cleaned, stamp, socket, name, int(epoch)),
    )
    if cursor.rowcount > 0:
        logger.info(
            "session_label_set_for_instance",
            tmux_socket=socket,
            tmux_name=name,
            tmux_created_epoch=int(epoch),
            note="title only; no identity column is in the SET clause",
        )
        return True
    logger.warning(
        "session_label_no_such_instance",
        tmux_socket=socket,
        tmux_name=name,
        tmux_created_epoch=int(epoch),
        note="no row carries that instance triple, so nothing was labelled",
    )
    return False

File: client/js/session-label.js
/**
 * THE ONE PLACE THAT DECIDES WHAT A SESSION IS CALLED ON SCREEN.
 *
 * A session's NAME IS A LABEL - free-form text a human typed, stored as
 * ``sessions.title`` and delivered as ``label``. The tmux session name is
 * an internal handle derived from that label once, at creation, and never
 * moved again. Keeping them separate is what stops a rename from moving
 * the field session identity is keyed on.
 *
 * WHY THIS IS A MODULE AND NOT A METHOD. Five surfaces render a session's
 * name: the launchpad rows, the browser tab title, the in-page header, a
 * toast card and the attribution prompt. Every one of them has to answer
 * the same question - what do I show when this session has no label? -
 * and a session with no label is not an edge case: it is EVERY session
 * that existed before labels did, plus every external session this app
 * never created. Three surfaces each answering that question separately
 * will drift, and the drift is invisible, because each one looks correct
 * on its own. So the rule lives here, once, and the surfaces call it.
 *
 * THE RULE HAS THREE OUTCOMES, NOT TWO:
 *
 *   1. a label      -> the label, verbatim. Whatever a human typed, and
 *                      that is the whole point of the feature: spaces,
 *                      ``:``, ``.``, quotes and ``$`` are all legal in a
 *                      label because a label is never handed to tmux.
 *   2. no label     -> the ``cloude_``-stripped tmux name. This is
 *                      EXACTLY what every one of these surfaces rendered
 *                      before labels existed, so a session with no label
 *                      looks precisely as it always did.
 *   3. neither      -> null, meaning THIS SESSION CANNOT BE NAMED. Never
 *                      an empty string, never the literal word "null",
 *                      never a silent blank - rendering nothing where a
 *                      name goes is worse than rendering the handle, and
 *                      far worse than saying you do not know.
 *
 * WHAT IS DELIBERATELY *NOT* HERE. Outcome 3 is returned, not rendered.
 * The surfaces word it differently on purpose - the tab title falls back
 * to the bare brand, a toast card says so in a sentence - and that is a
 * difference in WORDING, not a second fallback rule. A caller that
 * renders null as a blank has a bug; a caller that renders it as its own
 * sentence is correct.
 *
 * NOT AN IDENTITY, EITHER. Two sessions may carry the same label, and a
 * label matches nothing on the server. Anything that LOOKS UP a session -
 * the deep-link slug matcher, group membership, the adopt POST body, the
 * pinned-theme PATCH URL - must keep using the tmux name. This module is
 * for what a human reads and for nothing else.
 */
(function () {
    'use strict';

    /**
     * The prefix this app puts on the tmux names it creates. Stripped for
     * display because it is an artefact of the launcher, not anything a
     * user typed. Mirrors APP_TMUX_PREFIX in src/core/session_label.py.
     */
    var APP_TMUX_PREFIX = 'cloude_';

    /**
     * Longest label the SERVER will store. Mirrors LABEL_MAX_CHARS in
     * src/core/session_label.py.
     *
     * WHY A MIRROR AND NOT A FETCH. Nothing serves this constant, and an
     * editor has to know its own limit before it can draw a `maxlength`,
     * which is before any request it could learn the answer from. So it
     * is copied - but copied ONCE, here, rather than into each of the
     * three rename controls, and the copy is enforced:
     * tests/test_label_constants_parity.py parses both files and fails if
     * the two numbers drift. A mirror nobody checks is just a second
     * declaration waiting to disagree.
     *
     * The direction of a drift decides how it hurts, which is why the
     * parity test asserts EQUALITY rather than `client <= server`. A
     * client that permits MORE than the server accepts sends a label that
     * comes back rejected after the user typed it; a client that permits
     * LESS silently truncates at the `maxlength` and stores something the
     * user did not write, with no error anywhere. The second is worse and
     * is exactly what the hardcoded 64 was doing.
     * @type {number}
     */
    var LABEL_MAX_CHARS = 200;

    /**
     * Control characters, which no label may contain. Mirrors
     * _CONTROL_CHARS in src/core/session_label.py, newline and tab
     * included: a label is rendered on one line on every surface.
     * @type {RegExp}
     */
    var CONTROL_CHARS = /[\x00-\x1f\x7f]/;

    /**
     * What a surface shows for outcome 3. A sentence, because a blank
     * cell and an unknowable one look identical to a reader and mean
     * opposite things.
     */
    var UNKNOWN = 'unknown session';

    /**
     * Trim a value only if it is genuinely a string.
     *
     * Description: a JSON ``null`` arrives as null and a mis-shaped
     * payload could carry a number or an object. None of those is a
     * name, and ``String(null)`` would put the literal word "null" in a
     * browser tab, which is the exact failure this returns null for.
     * Inputs: value (*) - anything.
     * Output: string|null - the trimmed string, or null.
     */
    function cleanString(value) {
        if (typeof value !== 'string') return null;
        var trimmed = value.trim();
        return trimmed.length ? trimmed : null;
    }

    /**
     * Strip the app's tmux prefix and derive a readable display form.
     *
     * Description: THE ONE PLACE THIS APP TURNS A TMUX NAME INTO
     *   SOMETHING A HUMAN READS, mirroring ``label_from_tmux_name`` in
     *   ``src/core/session_label.py`` byte for byte on the transform
     *   itself - strip the ``cloude_`` prefix (only when something
     *   survives it), then replace every underscore with a space. The
     *   two used to disagree: this function stripped the prefix and
     *   stopped, so a session named ``Media_Compression`` rendered here
     *   with its underscore intact while the SAME name, run through the
     *   server's v9 label backfill, rendered as ``Media Compression`` -
     *   one conversation, two spellings, on two surfaces. Pinned equal
     *   by tests/test_label_derivation_parity.node.mjs (JS side) and
     *   tests/test_label_derivation_parity.py (Python side), both
     *   walking the SAME table of cases in tests/label_derivation_cases.json
     *   so the two can never drift apart again without a failing test.
     *
     *   ONE DELIBERATE NON-MIRROR. Python falls back to the ORIGINAL
     *   tmux name when the derived form is empty (``or tmux_name``), so
     *   ``label_from_tmux_name('cloude_')`` returns ``'cloude'`` rather
     *   than nothing - because that function feeds a stored ``title``
     *   column that is never allowed to go blank. This function instead
     *   returns null for that case, because null here means outcome 3
     *   of the module header's three-outcome rule - "this session
     *   cannot be named" - and a caller renders that as its own
     *   sentence (``UNKNOWN`` or similar) rather than a fabricated
     *   handle. Different callers, different empty-case contracts;
     *   the TRANSFORM applied to a non-degenerate name is identical.
     * Inputs: tmuxName (string|null).
     * Output: string|null - the display form, or null when nothing is
     *   left (a name that is ONLY the prefix strips to the empty string,
     *   which must not be rendered as a nameless row).
     * Example: SessionLabel.stripAppPrefix('cloude_Media')  -> 'Media'
     * Example: SessionLabel.stripAppPrefix('Media_Compression')
     *   -> 'Media Compression'
     */
    function stripAppPrefix(tmuxName) {
        var name = cleanString(tmuxName);
        if (name === null) return null;
        if (name.indexOf(APP_TMUX_PREFIX) === 0) {
            name = name.slice(APP_TMUX_PREFIX.length);
        }
        if (!name.length) return null;
        var derived = name.split('_').join(' ').trim();
        return derived.length ? derived : null;
    }

    /**
     * The string a HUMAN should see for one session, or null.
     *
     * Description: the whole fallback rule, and the only copy of it.
     *
     *   THE ONE PRESENTATION SWITCH, AND WHY IT IS NOT A SECOND RULE.
     *   ``stripPrefix: false`` keeps the ``cloude_`` prefix on the
     *   fallback. The FALLBACK CHAIN is identical either way - label,
     *   else the tmux name, else null - and only how the tmux name is
     *   spelled changes. The attribution prompt is the one surface that
     *   passes it, because there the prefix is EVIDENCE: that card is
     *   asking "did you start this session?", one of its own hints is
     *   that the name matches the auto-generated form, and the user may
     *   need to match the exact string against their own ``tmux ls``.
     *   Every other surface strips it, because there it is an artefact
     *   of the launcher and not part of the name.
     * Inputs: row (object|null) - reads ``label`` and ``name``. Either
     *   may be absent, empty or the wrong type. options (object|null) -
     *   ``{stripPrefix: false}`` to keep the ``cloude_`` prefix on the
     *   fallback; stripping is the default.
     * Output: string|null - see outcomes 1-3 in the module header. Never
     *   an empty string.
     * Example: SessionLabel.resolve({name: 'cloude_Media',
     *            label: 'Media Compression'})  -> 'Media Compression'
     */
    function resolve(row, options) {
        if (!row || typeof row !== 'object') return null;
        var label = cleanString(row.label);
        if (label !== null) return label;
        var strip = !(options && options.stripPrefix === false);
        return strip ? stripAppPrefix(row.name) : cleanString(row.name);
    }

    /**
     * The same rule, for a toast's field names.
     *
     * Description: a toast carries ``session_label`` / ``session_name``
     * rather than ``label`` / ``name``, because a toast already has a
     * ``title`` of its own and the two would collide. That is a SHAPE
     * difference, not a policy difference, so it is normalised into
     * :func:`resolve` rather than given a second copy of the chain.
     * Inputs: toast (object|null) - a server-shape toast.
     * Output: string|null - same three outcomes as :func:`resolve`.
     * Example: SessionLabel.resolveToast({session_name: 'cloude_a'}) -> 'a'
     */
    function resolveToast(toast) {
        if (!toast || typeof toast !== 'object') return null;
        return resolve({ label: toast.session_label, name: toast.session_name });
    }

    /**
     * What a rename editor should OPEN ON, read off the element that is
     * rendering the name.
     *
     * Description: all three rename controls seed themselves from the
     * RENDERED display value rather than re-deriving one. That is the
     * whole point: a seed computed a second time can disagree with what
     * the user is looking at, and when it did, a plain Enter overwrote a
     * user's label with a handle-derived string. Reading the element
     * means display and seed cannot disagree by construction - if a
     * surface's display half ever regresses, its seed regresses with it
     * and one test catches both.
     *
     * WHY `dataset.fullTitle` COMES FIRST. The in-page header
     * (`#header-title-text`) is MIDDLE-ELIDED by
     * client/js/header-title-fit.js: its `textContent` is a truncated
     * string with an ellipsis in it, and the fitter keeps the full value
     * in `dataset.fullTitle`, re-eliding from it on every resize. So on
     * that one surface the element holds its display value in two
     * places, and `textContent` is the WRONG one - seeding from it would
     * put "client: acme...$rate" in the box and store that truncation
     * over the real label on a plain Enter. That is the same silent data
     * loss this rule exists to remove, so the rule has to know about it.
     *
     * An element that does not elide carries no `fullTitle` and falls
     * through to its text, so this is one rule and not a special case.
     * Inputs: el (Element|null) - the element rendering the name.
     * Output: string - the seed, trimmed. Empty string when there is
     *   nothing to seed from, which a caller must treat as "do not open
     *   an editor" rather than as an empty starting value.
     * Example: SessionLabel.seedFromElement(titleEl)  -> 'Media Compression'
     */
    function seedFromElement(el) {
        if (!el) return '';
        var full = el.dataset ? el.dataset.fullTitle : null;
        var value = (typeof full === 'string' && full.length)
            ? full
            : (el.textContent || '');
        return String(value).trim();
    }

    /**
     * The server's label rule, mirrored so an obviously bad label is
     * refused without a round trip. THE SERVER REMAINS AUTHORITATIVE -
     * this is an early out, never the decision.
     *
     * WHAT IT DELIBERATELY DOES NOT REFUSE. The rename controls used to
     * enforce ``^[A-Za-z0-9_-]{1,64}$``, the old TMUX NAME rule, against
     * a field that is no longer a tmux name. That regex refused "Media
     * Compression" - the feature's own worked example - before it could
     * reach a server that would have accepted it happily. Spaces, ``:``,
     * ``.``, quotes, ``$`` and non-ASCII are all legal in a label,
     * because a label is never handed to tmux. Only the server's own two
     * refusals are mirrored: empty, and control characters.
     *
     * Inputs: raw (string|null) - the value from an edit control.
     * Output: object - ``{ok: true, value: string}`` with surrounding
     *   whitespace stripped, or ``{ok: false, reason: string}`` carrying
     *   a sentence fit to show the user.
     * Example: SessionLabel.validate(' Media Compression ')
     *   // {ok: true, value: 'Media Compression'}
     */
    function validate(raw) {
        var cleaned = cleanString(raw);
        if (cleaned === null) {
            return { ok: false, reason: 'a session label cannot be empty' };
        }
        if (cleaned.length > LABEL_MAX_CHARS) {
            return {
                ok: false,
                reason: 'a session label may be at most '
                    + LABEL_MAX_CHARS + ' characters',
            };
        }
        if (CONTROL_CHARS.test(cleaned)) {
            return {
                ok: false,
                reason: 'a session label cannot contain control characters'
                    + ' such as a newline or a tab',
            };
        }
        return { ok: true, value: cleaned };
    }

    window.SessionLabel = {
        APP_TMUX_PREFIX: APP_TMUX_PREFIX,
        LABEL_MAX_CHARS: LABEL_MAX_CHARS,
        UNKNOWN: UNKNOWN,
        stripAppPrefix: stripAppPrefix,
        resolve: resolve,
        resolveToast: resolveToast,
        seedFromElement: seedFromElement,
        validate: validate,
    };
})();
